pkg/cloudprovider/cloudprovider.go

```diff
@@ -74,12 +76,20 @@ type CloudProvider struct {
 }
 
 func New(ctx awscontext.Context) *CloudProvider {
+	clusterEndpoint, err := resolveClusterEndpoint(ctx, eks.New(ctx.Session))
+	if err != nil {
+		logging.FromContext(ctx).Debugf("unable to detect the cluster endpoint, %s", err)
+	} else {
+		logging.FromContext(ctx).With("cluster-endpoint", clusterEndpoint).Debugf("discovered cluster endpoint")
+		settings.FromContext(ctx).ClusterEndpoint = clusterEndpoint
```

Comment:
I don't think we should mutate this object directly. This feels wrong. Settings should be read-only